expression/builtin_time.go

```diff
@@ -2039,69 +2041,102 @@ type timestampFunctionClass struct {
+	} else {
+		sig = &builtinTimestamp1ArgSig{baseTimeBuiltinFunc{bf}}
+	}
 	return sig.setSelf(sig), nil
 }
 
-type builtinTimestampSig struct {
-	baseBuiltinFunc
+type builtinTimestamp1ArgSig struct {
+	baseTimeBuiltinFunc
 }
 
-// eval evals a builtinTimestampSig.
+// evalTime evals a builtinTimestamp1ArgSig.
 // See https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_timestamp
-func (b *builtinTimestampSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
-	if err != nil {
-		return d, errors.Trace(err)
-	}
-	if args[0].IsNull() {
-		return
+func (b *builtinTimestamp1ArgSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	s, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return types.Time{}, isNull, errors.Trace(err)
 	}
-	var arg0 types.Time
-	switch tp := args[0].Kind(); tp {
-	case types.KindInt64, types.KindUint64:
-		arg0, err = types.ParseDatetimeFromNum(args[0].GetInt64())
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-	case types.KindString, types.KindBytes, types.KindMysqlDecimal, types.KindFloat32, types.KindFloat64:
-		s, err1 := args[0].ToString()
-		if err1 != nil {
-			return d, errors.Trace(err1)
-		}
-		arg0, err = types.ParseTime(s, mysql.TypeDatetime, getFsp(s))
-		if err != nil {
-			return d, errors.Trace(err)
-		}
-	case types.KindMysqlTime:
-		arg0 = args[0].GetMysqlTime()
-	default:
-		return d, errors.Errorf("Unknown args type for timestamp %d", tp)
+	tm, err := types.ParseTime(s, mysql.TypeDatetime, getFsp(s))
+	if err != nil {
+		return types.Time{}, true, errors.Trace(handleInvalidTimeError(b.ctx, err))
 	}
-	if len(args) == 1 {
-		d.SetMysqlTime(arg0)
-		return
+	return tm, false, nil
+}
+
+type builtinTimestamp2ArgsSig struct {
+	baseTimeBuiltinFunc
+}
+
+// evalTime evals a builtinTimestamp2ArgsSig.
+// See https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_timestamp
+func (b *builtinTimestamp2ArgsSig) evalTime(row []types.Datum) (types.Time, bool, error) {
+	ctx := b.ctx.GetSessionVars().StmtCtx
+	arg0, isNull, err := b.args[0].EvalString(row, ctx)
+	if isNull || err != nil {
+		return types.Time{}, isNull, errors.Trace(err)
 	}
-	// If exists args[1].
-	s, err := args[1].ToString()
+	tm, err := types.ParseTime(arg0, mysql.TypeDatetime, getFsp(arg0))
 	if err != nil {
-		return d, errors.Trace(err)
+		return types.Time{}, true, errors.Trace(handleInvalidTimeError(b.ctx, err))
 	}
-	arg1, err := types.ParseDuration(s, getFsp(s))
-	if err != nil {
-		return d, errors.Trace(err)
+	arg1, isNull, err := b.args[1].EvalString(row, ctx)
+	if isNull || err != nil {
+		return types.Time{}, isNull, errors.Trace(err)
 	}
-	tmpDuration := arg0.Add(arg1)
-	result, err := tmpDuration.ConvertToTime(arg0.Type)
+	if !isDuration(arg1) {
+		return types.Time{}, true, nil
+	}
+    duration, err := types.ParseDuration(arg1, getFsp(arg1))
```

Comment:
how about add a cast and let child 1 return duration, so we can call `b.args[1].EvalDuration(row, ctx)`

Reply:
ditto